projects/gsc/experiments/sampled_kwinners/sampled_kwinner_le_sparse_net.py

```diff
@@ -0,0 +1,284 @@
+            k_inference_factor=k_inference_factor,
+            temperature=temperature,
+            eval_temperature=eval_temperature,
+            temperature_decay_rate=temperature_decay_rate,
+            relu=False
+        )
+        network.add_module("cnn{}_kwinner".format(suffix), kwinner)
+
+    if activation_fct_before_max_pool:
+        maxpool = nn.MaxPool2d(kernel_size=2)
+        network.add_module("cnn{}_maxpool".format(suffix), maxpool)
+
+
+def add_sparse_linear_layer(
+    network,
+    suffix,
+    input_size,
+    linear_n,
+    dropout,
+    use_batch_norm,
+    weight_sparsity,
+    percent_on,
+    k_inference_factor,
+    temperature,
+    eval_temperature,
+    temperature_decay_rate,
+    consolidated_sparse_weights,
+):
+    """Add sparse linear layer to network.
+
+    :param network: The network to add the sparse layer to
+    :param suffix: Layer suffix. Used to name its components
+    :param input_size: Input size
+    :param linear_n: Number of units
+    :param dropout: dropout value
+    :param use_batch_norm: whether or not to use batch norm
+    :param weight_sparsity: Pct of weights that are allowed to be non-zero
+    :param percent_on: Pct of ON (non-zero) units
+    :param k_inference_factor: During inference we increase percent_on by this factor
+    :param temperature: temperature to use when computing softmax in SampledKWinners.
+    :param eval_temperature:
+        temperature to use when computing softmax during evaluation in SampledKWinners.
+    :param temperature_decay_rate:
+        amount to decrease the temperature after each epoch of training.
+
+    """
+    linear = nn.Linear(input_size, linear_n)
+    if 0 < weight_sparsity < 1.0:
+        if consolidated_sparse_weights:
+            network.add_module(
+                "linear{}".format(suffix),
+                ConsolidatedSparseWeights(linear, weight_sparsity)
+            )
+        else:
+            network.add_module(
+                "linear{}".format(suffix), SparseWeights(linear, weight_sparsity)
+            )
+    else:
+        network.add_module("linear{}_linear".format(suffix), linear)
+
+    if use_batch_norm:
+        network.add_module("linear{}_bn".format(suffix),
+                           nn.BatchNorm1d(linear_n, affine=False))
+
+    if 0 < percent_on < 1.0:
+        network.add_module(
+            "linear{}_kwinners".format(suffix),
+            SampledKWinners(
+                percent_on=percent_on,
+                k_inference_factor=k_inference_factor,
+                temperature=temperature,
+                eval_temperature=eval_temperature,
+                temperature_decay_rate=temperature_decay_rate,
+                relu=False
+            ),
+        )
+    else:
+        network.add_module("linear{}_relu".format(suffix), nn.ReLU())
+
+    if dropout > 0.0:
+        network.add_module("linear{}_dropout".format(suffix), nn.Dropout(dropout))
+
+
+class SampledKWinnerLeSparseNet(nn.Sequential):
```

Comment:
Thanks for adding this. Originally, I didn't think we'd need a LeSparseNet, but it turns out this will be the model for our main baseline.